File: tests/test_phase7_signal_admission.cpp
// Phase 7 signal admission, from the dispatcher's side.
//
// This file previously asserted the opposite: that every signal method was
// refused in production. That was correct while the feature had never been
// trialled in a production-configuration binary. It has now been, on Godot
// 4.5.1, 4.6.2 and 4.7.2, so the contract these assert is the new one:
// signals dispatch, and the failure-injection seam does not exist.

#include "didi/gdextension/editor_hook.hpp"

#include <functional>
#include <stdexcept>
#include <string>

#define ASSERT_TRUE(cond) if (!(cond)) throw std::runtime_error("Assertion failed: " #cond)
#define ASSERT_FALSE(cond) ASSERT_TRUE(!(cond))
#define ASSERT_EQ(a, b) ASSERT_TRUE((a) == (b))
void registerTest(const std::string& name, std::function<void()> fn);

namespace {

// Admitted means reaching the bridge. Without an engine the bridge cannot
// satisfy the call, but the one answer that must no longer appear is the 501
// that said the method has no implementation at all.
void phase7_signal_production_admission_open() {
    auto& hook = didi::godot::EditorHook::instance();
    didi::godot::EditorHookTestAccess::setSessionKind(hook, didi::runtime::SessionKind::editor);

    for (const std::string method : {
             "signal.listConnections",
             "signal.connect",
             "signal.disconnect",
             "signal.emit",
         }) {
        const auto result = didi::godot::EditorHookTestAccess::executeOnMainThread(
            hook, method, didi::json::object());
        if (!result.contains("error")) continue;
        const auto code = result.at("error").at("code").get<int>();
        ASSERT_TRUE(code != 501);
        ASSERT_TRUE(code != 404);
    }
}

// The seam configurator is compiled out of every build except the dedicated
// signal-test extension. The native test binary is not that build, so the
// method must be unknown here -- the same answer a release would give.
void phase7_signal_test_seam_absent_from_production() {
    auto& hook = didi::godot::EditorHook::instance();
    didi::godot::EditorHookTestAccess::setSessionKind(hook, didi::runtime::SessionKind::editor);
    const auto result = didi::godot::EditorHookTestAccess::executeOnMainThread(
        hook, "phase7SignalTest.configure", {{"seam", "missing_required_api"}});
    ASSERT_TRUE(result.contains("error"));
    ASSERT_EQ(result.at("error").at("code").get<int>(), 404);
}

struct RegisterPhase7SignalAdmission {
    RegisterPhase7SignalAdmission() {
        registerTest("Phase7Signals.ProductionAdmissionOpen",
                     [] { phase7_signal_production_admission_open(); });
        registerTest("Phase7Signals.TestSeamAbsentFromProduction",
                     [] { phase7_signal_test_seam_absent_from_production(); });
    }
} register_phase7_signal_admission;

} // namespace

File: tests/test_phase7_signal_admission.py
"""The signal bridge is admitted to production; its test seams are not.

One macro used to control both, so the feature could not be admitted without
also compiling failure-injection seams into a shipping binary. These assert the
two are now separate, and that the seam configurator stays behind the gate --
that is the property a release depends on.
"""

from pathlib import Path
import unittest


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "src" / "gdextension" / "editor_hook.cpp"
SIGNAL_METHODS = (
    '"signal.listConnections"',
    '"signal.connect"',
    '"signal.disconnect"',
    '"signal.emit"',
)
SEAM_GUARD = "#if defined(DIDI_PHASE7_SIGNAL_TEST_SEAMS)"
SEAM_CONFIGURATOR = '"phase7SignalTest.configure"'


def _live_bridge_set(source: str) -> str:
    start = source.index("static const std::unordered_set<std::string> live_bridge_methods")
    return source[start:source.index("};", start)]


class Phase7SignalAdmissionSourceTests(unittest.TestCase):
    def setUp(self):
        self.source = SOURCE.read_text(encoding="utf-8")

    def test_signal_methods_are_admitted_unconditionally(self):
        # They must sit before the seam guard, or admission would once again
        # depend on compiling test seams into the binary users run.
        live_set = _live_bridge_set(self.source)
        production = live_set.split(SEAM_GUARD, 1)[0]
        for method in SIGNAL_METHODS:
            self.assertIn(method, production)

    def test_the_seam_configurator_never_reaches_production(self):
        # The safety-critical half. A shipping binary that can be told to force
        # its own failure paths is not a shipping binary.
        live_set = _live_bridge_set(self.source)
        self.assertIn(SEAM_GUARD, live_set)
        production, seam_only = live_set.split(SEAM_GUARD, 1)
        self.assertNotIn(SEAM_CONFIGURATOR, production)
        self.assertIn(SEAM_CONFIGURATOR, seam_only)

    def test_signal_methods_are_no_longer_listed_as_unimplemented(self):
        start = self.source.index(
            "static const std::unordered_set<std::string> registered_but_unimplemented"
        )
        compatibility_set = self.source[start:self.source.index("};", start)]
        for method in SIGNAL_METHODS:
            self.assertNotIn(method, compatibility_set)


if __name__ == "__main__":
    unittest.main()
